xdrjson/conversion.go

```
@@ -16,55 +16,18 @@ package xdrjson
 import "C"
 
 import (
-	"encoding"
 	"encoding/json"
-	"reflect"
 	"unsafe"
 
 	"github.com/pkg/errors"
 )
 
-// ConvertBytes takes an XDR object (`xdr`) and its serialized bytes (`field`)
-// and returns the raw JSON-formatted serialization of that object.
-// It can be unmarshalled to a proper JSON structure, but the raw bytes are
-// returned to avoid unnecessary round-trips. If there is an
-// error, it returns an empty string.
-//
-// The `xdr` object does not need to actually be initialized/valid:
-// we only use it to determine the name of the structure. We could just
-// accept a string, but that would make mistakes likelier than passing the
-// structure itself (by reference).
-func ConvertBytes(xdr interface{}, field []byte) (json.RawMessage, error) {
-	if len(field) == 0 {
-		return []byte(""), nil
-	}
-
-	xdrTypeName := reflect.TypeOf(xdr).Name()
-	return convertAnyBytes(xdrTypeName, field)
-}
-
-// ConvertInterface takes a valid XDR object (`xdr`) and returns
-// the raw JSON-formatted serialization of that object. If there is an
-// error, it returns an empty string.
-//
-// Unlike `ConvertBytes`, the value here needs to be valid and
-// serializable.
-func ConvertInterface(xdr encoding.BinaryMarshaler) (json.RawMessage, error) {
-	xdrTypeName := reflect.TypeOf(xdr).Name()
-	data, err := xdr.MarshalBinary()
-	if err != nil {
-		return []byte(""), errors.Wrapf(err, "failed to serialize XDR type '%s'", xdrTypeName)
-	}
-
-	return convertAnyBytes(xdrTypeName, data)
-}
```

`ConvertBytes` and `ConvertInterface` behave differently:
- `ConvertBytes` returns an empty string (`""`) if the `[]byte` is length zero.
- `ConvertInterface` returns an error if the `[]byte` is length zero, because the XDR decoder will error in that case for all XDR types.

@stellar/platform-eng Which is the correct behaviour? I think one must be a bug since both provide a different interface into the same capabilities.

There's a test at least that expects `ConvertBytes` to return an empty string, so seems some thought went into that use case.

Imo it'd be better if the function returned an error, because any other XDR decoder would return an error in this case, and to not return an error is silently hiding what is probably a problem. A system using this library can always translate the error into an empty string if it wants to hide errors.

Thoughts?

---

Both are correct. If you explicitly pass an empty byte array, it is the correct behavior to return an empty string. 

For example, you want to decode some byte array under the presumption that it is of a particular type. However, that byte array is empty because it is optional (for example, the restoration preamble is empty when nothing needs restoration). In that case, it's valid to just return an empty string, which will then `omitempty` later.

---

I don't think any other XDR decoder fails silently like that, but I'll have a look at the other decoders to check and see what they do.

I understand the design here was influenced by it's use directly in an application, but hiding a silent failure in a lib is not ideal because it's surprising.

I think if we move ahead with returning an error here, and change the PR so that the error is unique, then existing consumers that prefer to silently hide the error can do so via checking. I can look into validating this assumption before making a decision here.

---

The Rust and JS impls fail on empty strings, which I think is the intuitive thing to do here.

It also looks like 4 of the 9 call sites where the stellar-rpc repo uses the `ConvertBytes` function already check if the length is zero before calling the function. The remaining locations look trivial to change to do the same, and doing so would make the calling code more explicit about how it handles the empty case. So it looks like a net benefit to change to fail on empty.

I've left the Decode function to fail on empty, and the deprecated ConvertBytes function to return "" on empty.